src/components/DropMenu/DropMenu.jsx

```diff
@@ -0,0 +1,468 @@
+			onFocusOption
+		} = this.props;
+
+		this.setState({
+			isMouseTriggered: true
+		});
+
+		if (!optionProps.isDisabled && focusedIndex !== optionIndex) {
+			onFocusOption(optionIndex);
+		}
+	},
+
+	handleSelectOption(optionIndex, optionProps) {
+		const {
+			onSelect
+		} = this.props;
+
+		if (!optionProps.isDisabled) {
+			onSelect(optionIndex);
+		}
+	},
+
+	renderOption(optionProps, optionIndex, isGrouped) {
+		const {
+			selectedIndices,
+			focusedIndex,
+		} = this.props;
+
+		const {
+			isMouseTriggered
+		} = this.state;
+
+		const {
+			isDisabled
+		} = optionProps;
+
+		const isFocused = optionIndex === focusedIndex;
+		const isSelected = _.includes(selectedIndices, optionIndex);
+
+		return (
+			<div
+				className={boundClassNames(
+					'&-Option', {
+					'&-Option-is-grouped': isGrouped,
+					'&-Option-is-focused': isFocused,
+					'&-Option-is-selected': isSelected,
+					'&-Option-is-disabled': isDisabled,
+					'&-Option-is-null': _.isNull(optionIndex)
+				})}
+				onMouseMove={() => this.handleMouseFocusOption(optionIndex, optionProps)}
+				onClick={() => this.handleSelectOption(optionIndex, optionProps)}
+				ref={(optionDOMNode)=> {
+					if (isFocused && !isMouseTriggered) {
+						scrollParentTo(optionDOMNode);
+					}
+				}}
+				key={'DropMenuOption' + optionIndex}
+			>
+				{optionProps.children}
+			</div>
+		);
+	},
+
+	componentWillMount() {
+		// preprocess the options data before rendering
+		this.setState(this.getPreprocessedOptionData(this.props));
+	},
+
+	componentWillReceiveProps(nextProps) {
+		// only preprocess options data when it changes (via new props) - better performance than doing this each render
+		this.setState(this.getPreprocessedOptionData(nextProps));
+	},
+
+	render() {
+		const {
+			className,
+			style,
+			isDisabled,
+			isExpanded,
+			direction,
+			portalId,
+			onCollapse
+		} = this.props;
+
+		const {
+			optionGroups,
+			ungroupedOptionData,
+			optionGroupDataLookup,
+			nullOptions
+		} = this.state;
+
+		const controlProps = _.first(DropMenu.Control.findInAllAsProps(this.props));
+
+		return (
+			<div className={boundClassNames('&', '&-base', {
+				'&-is-expanded': isExpanded,
+				'&-direction-down': isExpanded && direction === 'down',
+				'&-direction-up': isExpanded && direction === 'up'
+			}, className)} style={style}>
+				<ContextMenu portalId={portalId} isExpanded={isExpanded} direction={direction} onChangeBounds={this.handleChangeBounds} onClickOut={onCollapse}>
+					<ContextMenu.Target>
+						<div
+							{...(!isDisabled ? {
+								tabIndex: 0,
+								onClick: this.handleClick,
+								onKeyDown: this.handleKeydown
+							} : null)}
+							{...controlProps}
+							className={boundClassNames('&-Control', _.get(controlProps, 'className'))}
+						/>
+					</ContextMenu.Target>
+					<ContextMenu.FlyOut className={boundClassNames('&', className)}>
+						{
+							_.map(nullOptions, (optionProps) => this.renderOption(optionProps, null))
```

Comment:
i agree generally, but the reason i create a large array of nodes is so that i can call the `joinArray` function to place dividers between each of the groups